Common CI script

waleko opened the pull request:
This is a pull request resolving issue #39.

New bash script `runchecks` replaces following scripts:

- `checkcommit`
- `runcheckendline`
- `runcheckmigration`
- `runflake8`
- `runisort`

It accepts `--skip-{check-name}` options (e.g. `./runchecks --skip-isort --skip-checkcommit`).
You can also run it with `--help` option (if you want to for some reason)

P.S. It's a GCI Task

dwang commented on runchecks:
```diff
@@ -0,0 +1,115 @@
+}
+
+runcheckmigration() {
+    set -e
```
You can move `set -e` to the top and it will apply for the entire script.

dwang commented on runchecks:
```diff
@@ -0,0 +1,115 @@
+}
+
+runflake8() {
+    set -e
```
You can move `set -e` to the top and it will apply for the entire script.

dwang commented on runchecks:
```diff
@@ -0,0 +1,115 @@
+}
+
+runisort() {
+    set -e
```
You can move `set -e` to the top and it will apply for the entire script.

dwang commented on runchecks:
```diff
@@ -0,0 +1,115 @@
+      COMMIT_MESSAGE=$(git log $TRAVIS_PULL_REQUEST_SHA --format=%B -n 1)
+      printf "Checking commit message:\n\n"
+      printf "$COMMIT_MESSAGE\n\n"
+      checkcommit --message "$COMMIT_MESSAGE"
```
The arguments aren't being passed correctly. Can you take a look?
https://travis-ci.org/openwisp/openwisp-utils/jobs/621280059#L424

waleko replied:
The thing with passing arguments turned up to be not the problem with my script. It's a problem with coverage. You can see some [old build](https://travis-ci.org/openwisp/openwisp-utils/jobs/525184823#L636) (even without `runchecks`) where these errors are still appearing. Presumably, `runchecks.py` runs `checkcommit` and `checkmigration` functions, that have no arguments to process which results in an error. I can take a deeper look into it if you wish.

dwang commented on runchecks:
```diff
@@ -0,0 +1,115 @@
+
+runcheckmigration() {
+    set -e
+    checkmigrations --migration-path ./tests/test_project/migrations/ || exit 1
```
The arguments aren't being passed correctly. Can you take a look?
https://travis-ci.org/openwisp/openwisp-utils/jobs/621280059#L427

waleko replied:
The thing with passing arguments turned up to be not the problem with my script. It's a problem with coverage. You can see some [old build](https://travis-ci.org/openwisp/openwisp-utils/jobs/525184823#L636) (even without `runchecks`) where these errors are still appearing. Presumably, `runchecks.py` runs `checkcommit` and `checkmigration` functions, that have no arguments to process which results in an error. I can take a deeper look into it if you wish.

waleko commented:
Hi, thanks for suggestions!

I've moved `set -e` to the main level. I also squashed my commits in one.

The thing with passing arguments turned up to be not the problem with my script. It's a problem with coverage. You can see some [old build](https://travis-ci.org/openwisp/openwisp-utils/jobs/525184823#L636) (even without `runchecks`) where these errors are still appearing. Presumably, `runchecks.py` runs `checkcommit` and `checkmigration` functions, that have no arguments to process which results in an error. I can take a deeper look into it if you wish.

If the check is failed the script will return non-zero exit code as you can see in [this build](https://travis-ci.org/openwisp/openwisp-utils/jobs/621277720#L426) when I myself ran into an issue of not following contributing guidelines on commit names (duh).

atb00ker commented:
Great work here! 

> @atb00ker thoughts about this?

I'd defer this to @nemesisdesign 

> Presumably, runchecks.py runs checkcommit and checkmigration functions, that have no arguments to process which results in an error. I can take a deeper look into it if you wish.

Yes, it's just the output while testing the scripts. :smile:


Is this ready for testing, shall I try it?

atb00ker commented on runchecks:
```diff
@@ -0,0 +1,114 @@
+    runflake8
+fi
+
+if ! $skip_checkisort; then
```
This is not correct. Should be `$skip_isort` according to the variable above.

atb00ker commented on runchecks:
```diff
@@ -0,0 +1,114 @@
+#!/bin/bash
+
+show_help() {
```
Just for the sake of prettification of the output, can you please try:
```
    printf "Usage:\n"
    printf "runchecks [--help] [--skip-checkendline] [--skip-checkmigration] [--skip-flake8] [--skip-isort] [--skip-checkcommit]\n"
    printf "\n\n"
    printf "Options:\n"
    printf "  --help\t\t\t: Show this help text\n"
    printf "  --skip-checkendline\t\t: Skip blank endline check\n"
    printf "  --skip-checkmigration\t\t: Skip migration name check\n"
    printf "  --skip-flake8\t\t\t: Skip flake8 check\n"
    printf "  --skip-isort\t\t\t: Skip isort check\n"
    printf "  --skip-checkcommit\t\t: Skip commit check\n"
```

waleko commented:
@atb00ker Made the changes

atb00ker commented:
Okay, I just realized a big mistake in the implementation.

The primary reason of of making this change is that in any repository we can simply install `openwisp-utils` and get our set quality assurance checks but I just realized that's not something we can do with shell script! (If you have a possible solution, please let me know, we can discuss it.)

Hence, I think you'll need to write this shell script in python instead as done for `checkcommit` here:
- https://github.com/openwisp/openwisp-utils/blob/master/openwisp_utils/qa.py#L74
- https://github.com/openwisp/openwisp-utils/blob/master/setup.py#L37

Now, we can install `openwisp-utils` and use `checkcommit` command to get our work done in repository.

waleko commented:
@atb00ker @dwang Thankfully, I found a much simpler workaround. I just added the script to `setup.py`.
```python
#... 
entry_points={
        'console_scripts': [
            'checkmigrations = openwisp_utils.qa:check_migration_name',
            'checkcommit = openwisp_utils.qa:check_commit_message',
        ],
    },
scripts=["runchecks"],
include_package_data=True,
#... 
```
Now when we call `pip install openwisp-utils` runchecks is installed automatically :smile:.
I think, it's a much simpler solution than rewriting it to python, as many of the checks themselves use command line, so with python it would be just a bit easier to write help page and to parse the arguments.

I also added `--migration-path <path>` argument to `runchecks`, as discussed with @atb00ker, and tweaked exit codes to be different for different checks and added `echo "Running {...} check"`, so it would be easier to understand which check failed.

dwang commented:
I'm okay with this, are you @atb00ker?
I'll do some testing right now and report back.

atb00ker commented:
@waleko That sounds great. It'll I'll try to test it later. For now, please try testing it:
- Install your openwisp_utils in another openwisp repository and try to make sure all the tests run with the command.

I have a couple of question:
1. If in future we decide to make this shell script as python script, would we need to change anything in repositories that it running on. (Example `django-loci` is going to use these these test, so, we will run `runchecks` in loci, but if we change it to python script, does anything change for `django-loci` repository to run the script)
2. Just confirming but it sounds like `--migration-path` is the only option you think is required from the old script but as a matter of fact we need other options like `quite`, `migrations_to_ignore` etc. Hope I didn't convey that incorrectly, `--migration-path` was only an example but we don't want to remove any of the features of the old scripts because of this new one! :smile:

atb00ker commented:
@dwang Sounds great, thanks! :smile:

waleko commented:
@dwang @atb00ker I hard-coded all parameters used by `checkcommit` and `checkmigrations`. I also "redisgned" help page, so it's easier to understand what parameters are for which check. 
```
Usage:
  runchecks --help
  runchecks {--migration-path <path> [--migrations-to-ignore <int>] | --skip-checkmigrations}
            {[--message <commit-message>] | --skip-checkcommit}
            [--skip-checkendline]
            [--skip-flake8]
            [--skip-isort]
            [--quiet]

Runs all checks for quality assurance. Unneeded checks may be skipped by passing --skip-<check-name> argument.

General options:
  --help                        : Show this help text
  --quiet                       : Suppress output
Migration check options:
  --migration-path <path>       : Path to migrations/ folder
  --migrations-to-ignore <int>  : Number of migrations after which checking of migration file names should begin,
                                  say, if checking needs to start after 0003_auto_20150410_3242.py value should be 3
  --skip-checkmigrations        : Skip migration name check
Commit name check options:
  --message <commit-message>    : Commit message (by default is already equal to $COMMIT_MESSAGE)
  --skip-checkcommit            : Skip commit check
Blank endlines check options:
  --skip-checkendline           : Skip blank endlines check
Flake8 check options:
  --skip-flake8                 : Skip flake8 check
Isort check options:
  --skip-isort                  : Skip isort check
```
Now if we change `runchecks` to a python script, all we would need to do is to parse all arguments and pass the arguments object to an auxiliary function and it will just use its arguments 😄. The only limitation to it is that all arguments names have to be unique, but I think, that it's still a good thing, because it would be better having a more descriptive argument name.

I also added a `--quiet` parameter. It suppresses `runcheck` output and output of all auxiliary programs as well.

nemesisdesign commented:
@dwang 

> I also noticed that all the checks are run at the beginning now, but maybe it might be better to run them all at the end since code changes will not be tested if something minor like the commit message check fails. @nemesisdesign @atb00ker thoughts about this?

If something is wrong, it is better to fail as early as possible so that who's working on the PR finds out early and can fix quickly.

nemesisdesign commented on runchecks:
```diff
@@ -0,0 +1,188 @@
+#!/bin/bash
+
+show_help() {
+  printf "Usage:\n"
```
when the final `\n` is needed, why can't you just use `echo` instead? I think that is cleaner. Unless you have a better reason to use `printf` I ask you to use `echo`.

waleko replied:
`echo` doesn't work so well with `\t`. Sometimes it just prints "\t" in plaintext, so I changed it to `printf`, so it has a more stable behavior.

nemesisdesign commented on .travis.yml:
```diff
@@ -32,21 +32,10 @@ install:
-  - ./runisort
-  - ./runcheckmigration
-  - ./runcheckendline
+  - runchecks --migration-path "./tests/test_project/migrations"
```
let's rename this script: `openwisp-utils-qa-checks`, this way we will make it obvious where the script comes from and contributors will quickly be able to find it

nemesisdesign commented on runchecks:
```diff
@@ -0,0 +1,188 @@
+
+## Main
+
+skip_checkendline=false
```
please let's stay consistent and use uppercase name for these variables

nemesisdesign commented on setup.py:
```diff
@@ -37,6 +37,7 @@
             'checkcommit = openwisp_utils.qa:check_commit_message',
         ],
     },
+    scripts=["openwisp-utils-qa-checks"],
```
please use single quotes (`'`) for consistency, keep in mind to always be consistent with the styles already used in the program
